analysis/exponential.lean

```diff
@@ -0,0 +1,921 @@
+have sin (arcsin x) ^ 2 + cos (arcsin x) ^ 2 = 1 := sin_pow_two_add_cos_pow_two (arcsin x),
+begin
+  rw [← eq_sub_iff_add_eq', ← sqrt_inj (pow_two_nonneg _) (sub_nonneg.2 (sin_pow_two_le_one (arcsin x))),
+    pow_two, sqrt_mul_self (cos_arcsin_nonneg _)] at this,
+  rw [this, sin_arcsin hx₁ hx₂],
+end
+
+lemma sin_arccos {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : sin (arccos x) = sqrt (1 - x ^ 2) :=
+by rw [arccos_eq_arcsin_add_pi_div_two, sin_pi_div_two_sub, cos_arcsin hx₁ hx₂]
+
+lemma abs_div_sqrt_one_add_lt (x : ℝ) : abs (x / sqrt (1 + x ^ 2)) < 1 :=
+have h₁ : 0 < 1 + x ^ 2, from add_pos_of_pos_of_nonneg zero_lt_one (pow_two_nonneg _),
+have h₂ : 0 < sqrt (1 + x ^ 2), from sqrt_pos.2 h₁,
+by rw [abs_div, div_lt_iff (abs_pos_of_pos h₂), one_mul,
+    mul_self_lt_mul_self_iff (abs_nonneg x) (abs_nonneg _),
+    ← abs_mul, ← abs_mul, mul_self_sqrt (add_nonneg zero_le_one (pow_two_nonneg _)),
+    abs_of_nonneg (mul_self_nonneg x), abs_of_nonneg (le_of_lt h₁), pow_two, add_comm];
+  exact lt_add_one _
+
+lemma div_sqrt_one_add_lt_one (x : ℝ) : x / sqrt (1 + x ^ 2) < 1 :=
+(abs_lt.1 (abs_div_sqrt_one_add_lt _)).2
+
+lemma neg_one_lt_div_sqrt_one_add (x : ℝ) : -1 < x / sqrt (1 + x ^ 2) :=
+(abs_lt.1 (abs_div_sqrt_one_add_lt _)).1
+
+lemma tan_pos_of_pos_of_lt_pi_div_two {x : ℝ} (h0x : 0 < x) (hxp : x < π / 2) : 0 < tan x :=
+by rw tan_eq_sin_div_cos; exact div_pos (sin_pos_of_pos_of_lt_pi h0x (by linarith))
+  (cos_pos_of_neg_pi_div_two_lt_of_lt_pi_div_two (by linarith) hxp)
+
+lemma tan_nonneg_of_nonneg_of_le_pi_div_two {x : ℝ} (h0x : 0 ≤ x) (hxp : x ≤ π / 2) : 0 ≤ tan x :=
+(lt_or_eq_of_le h0x).elim
+  (λ hx0, (lt_or_eq_of_le hxp).elim
+    (le_of_lt ∘ tan_pos_of_pos_of_lt_pi_div_two hx0)
+    (λ h, by simp [h, tan_eq_sin_div_cos]))
+  (λ h, by simp [h.symm])
+
+lemma tan_neg_of_neg_of_pi_div_two_lt {x : ℝ} (hx0 : x < 0) (hpx : -(π / 2) < x) : tan x < 0 :=
+neg_pos.1 (tan_neg x ▸ tan_pos_of_pos_of_lt_pi_div_two (by linarith) (by linarith using [pi_pos]))
+
+lemma tan_nonpos_of_nonpos_of_neg_pi_div_two_le {x : ℝ} (hx0 : x ≤ 0) (hpx : -(π / 2) ≤ x) : tan x ≤ 0 :=
+neg_nonneg.1 (tan_neg x ▸ tan_nonneg_of_nonneg_of_le_pi_div_two (by linarith) (by linarith using [pi_pos]))
+
+lemma tan_lt_tan_of_nonneg_of_lt_pi_div_two {x y : ℝ} (hx₁ : 0 ≤ x) (hx₂ : x < π / 2) (hy₁ : 0 ≤ y)
+  (hy₂ : y < π / 2) (hxy : x < y) : tan x < tan y :=
+begin
+  rw [tan_eq_sin_div_cos, tan_eq_sin_div_cos],
+  exact div_lt_div
+    (sin_lt_sin_of_le_of_le_pi_div_two (by linarith) (le_of_lt hx₂)
+      (by linarith) (le_of_lt hy₂) hxy)
+    (cos_le_cos_of_nonneg_of_le_pi hx₁ (by linarith) hy₁ (by linarith) (le_of_lt hxy))
+    (sin_nonneg_of_nonneg_of_le_pi hy₁ (by linarith))
+    (cos_pos_of_neg_pi_div_two_lt_of_lt_pi_div_two (by linarith) hy₂)
+end
+
+lemma tan_lt_tan_of_lt_of_lt_pi_div_two {x y : ℝ} (hx₁ : -(π / 2) < x) (hx₂ : x < π / 2)
+  (hy₁ : -(π / 2) < y) (hy₂ : y < π / 2) (hxy : x < y) : tan x < tan y :=
+(le_total x 0).elim

```

Comment:
Why not using `match le_total x 0, le_total y 0 with ...`. it looks like you have 4 cases anyway. And its more readable.